src/core/shred.zig

```diff
@@ -11,4 +12,56 @@ pub const ShredVersion = struct {
     pub fn init_manually_set(version: u16) Self {
         return Self{ .value = version };
     }
+
+    pub fn version_from_hash(hash: *Hash) u16 {
+        const hash_bytes = hash.data;
+        var accum: [2]u8 = .{ 0, 0 };
+        var chunks = std.mem.window(u8, &hash_bytes, 2, 2);
+
+        while (chunks.next()) |chunk| {
+            accum[0] ^= chunk[0];
+            accum[1] ^= chunk[1];
+        }
+        var version = (@as(u16, accum[0]) << 8) | accum[1];
+        return version +| 1;
+    }
+
+    pub fn compute_shred_version(alloc: Allocator, genesis_hash: Hash, maybe_hard_forks: ?HardForks) u16 {
+        var hash = genesis_hash;
+        if (maybe_hard_forks) |hard_forks| {
+            var buf: [16]u8 = undefined;
+            for (hard_forks.get_forks()) |hard_fork| {
+                std.mem.writeIntLittle(u64, buf[0..8], hard_fork.slot);
+                std.mem.writeIntLittle(u64, buf[8..], @as(u64, hard_fork.count));
+                hash = Hash.extend_and_hash(genesis_hash, &buf, alloc) catch genesis_hash;
```

Comment:
the only way this fcn fails (and catch is called) is if it runs out of memory, right? - in this case, we should propagate the out of memory error instead of catching

also, the first parameter should be `hash`

Reply:
I will add more tests to make sure the calculation is correct